blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,701 @@
+
+There are different approaches for letting the order service know the other two about new purchase orders;
+e.g. it could invoke some https://en.wikipedia.org/wiki/Representational_state_transfer[REST], https://grpc.io/[grpc] or other (synchronous) API provided by these services.
+This might create some undesired coupling, though: the sending service must know which other services to invoke and where to find them.
+It also must be prepared for these services temporarily not being available.
+Service meshes such as https://istio.io/[Istio] can come in helpful here, by providing capabilities like request routing, retries, circuit breakers and much more.
+
+The general issue of any synchronous approach is that one service cannot really function without the other services which it invokes.
+While buffering and retrying might help in cases where other services only need to be _notified_ of certain events,
+this is not the case if a service actually needs to _query_ other services for information.
+For instance, when a purchase order is placed, the order service might need to obtain the information how many times the purchased item is on stock from an inventory service.
+
+Another downside of such a synchronous approach is that it lacks re-playability,
+i.e. the possibility for new consumers to arrive long after events have been sent and still be able to consume the entire event stream from the beginning.
+
+Both problems can be addressed by using an asynchronous data exchange approach instead:
+i.e having the order, inventory and other services propagate events through a durable message log such as http://kafka.apache.org/[Apache Kafka].
+By subscribing to these event streams, each service will be notified about the data change of other services.
+It can react to these events and, if needed, create a local representation of that data in its own database,
+using a representation specifically tailored towards its own needs.
+For instance, such view might be denormalized to efficiently support specific access patterns, or it may only contain a subset of the original data that's relevant to the consuming service.
+
+Durable logs also support re-playability,
+i.e. new consumers can be added as needed, enabling use cases you might not have had in mind originally,
+and without touching the source service.
+E.g. consider a data warehouse which should keep information about all the orders ever placed, or some full-text search functionality on purchase orders based on https://www.elastic.co/products/elasticsearch[Elasticsearch].
+Once the purchase order events are in a Kafka topic
+(and depending on the topic's retention policy, they could remain there for an arbitrary long time, or even indefinitely),
+new consumers can subscribe, process the topic from the very beginning and materialize a view of all the data in a microservice's database, search index, data warehouse etc.
```

Comment:
I see; do you have an alternative for replayability? Also I'm curious, that thread you linked is nearly three years old, perhaps things have changed since then? I've seen references to long retention, but I'm sure what's the current agreed on thinking on that one is.

Reply:
while it is right that very long (indefinite) storage brings its own challenges there are real-world examples where this is done. the "most famous" one that quickly comes to my mind is the NYT approach towards permanent storage https://www.confluent.io/blog/publishing-apache-kafka-new-york-times/ and a referring post https://www.confluent.io/blog/okay-store-data-apache-kafka/

Reply:
From the NYT blog:
```
Because the topic is single-partition, it needs to be stored on a single disk, due to the way Kafka stores partitions. This is not a problem for us in practice, since all our content is text produced by humans — our total corpus right now is less than 100GB, and disks are growing bigger faster than our journalists can write.
```
So that's 100GB for > 150 years of data. I think many, many use cases customers in production will have much greater different scale than  .66 GB / year and design systems to use more than one partition. 

I'm not saying that the approach is invalid, I'm just saying that proposing this sort of architecture (indefinite storage in the Kafka cluster, no event storage in the DB) has serious operational implication and at minimal we should highlight this fact so readers are aware of trade off & alternative.

As to reply ability, I think the most likely case is going to after an outage where we've used traditional message queueing for ages. The idea of a whole new service replaying 5 years of message history from multiple services to create an initial state is not a daily or weekly requirement at most organizations. If the requirement does come up, I would prefer to leverage an event log in my RDBMS and create a one off debezium connector for this specific purpose. This gets in to the whole idea of optionality I write about [here](https://medium.com/@justinmichaelholmes/creating-optionality-with-domain-event-logging-7eb8f1f422e7)

Again - these alternatives. Neither is right or wrong and all I am asking is that we highlight this approach as one of many.